contracts/src/FunctionGatewayV2.sol

```diff
@@ -0,0 +1,165 @@
+    ) external payable returns (bytes32) {
+        requests[nonce++] = keccak256(FunctionRequest({
+            functionId: _functionId,
+            inputBytes: _input,
+            context: _context,
+            callbackAddress: msg.sender,
+            callbackSelector: _callbackSelector
+        }));
+        // TODO: emit event with preference
+    }
+
+    function fulfillCallbackRequestV1(
+        bytes32 _requestId,
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _output,
+        bytes memory _proof,
+        bytes memory _context,
+        address _callbackAddress,
+        bytes32 _callbackSelector
+    ) external {
+        bytes32 requestHash = keccak256(FunctionRequest({
+            functionId: _functionId,
+            inputBytes: _input,
+            context: _context,
+            callbackAddress: _callbackAddress,
+            callbackSelector: _callbackSelector
+        }));
+        require(requests[_requestId] == requestHash, "Request not found");
+
+        address verifier = functionIdToVerifier[_functionId];
+        bytes32 inputHash = sha256(_input);
+        bytes32 outputHash = sha256(_output);
+        if (
+            !IFunctionVerifier(verifier).verify(
+                inputHash,
+                outputHash,
+                _proof
+            )
+        ) {
+            revert InvalidProof();
+        }
+
+        (bool status, ) = _callbackAddress.call(
+            abi.encodeWithSelector(_callbackSelector, _input, _output, _context)
+        );
+        if (!status) {
+            revert CallbackFailed(_callbackAddress, _callbackSelector);
+        }
+
+        delete requests[_requestId];
+    }
+
+    function storeRequestV1(
+        bytes32 _functionId,
+        bytes memory _input
+    ) {
+        emit StoreRequest(_functionId, _input, uint256(0), address(0), bytes4(0));
+    }
+
+    function storeRequestV1(
+        bytes32 _functionId,
+        bytes memory _input,
+        uint256 _chain_id,
+        address _address,
+        bytes4 _selector,
+        uint8 preference // Slow, medium, fast, aggregation or not, etc.
+    ) {
+        emit StoreRequest(_functionId, _input, _chain_id, _address, _selector);
+    }
+
+    function fulfillStoreRequest(
+        bytes32 _functionId,
+        bytes memory _input,
+        bytes memory _output,
+        bytes memory _proof,
+        address _address,
+        bytes4 _selector
+    ) {
+        address verifier = functionIdToVerifier[_functionId];
+        bytes32 inputHash = sha256(_input);
+        bytes32 outputHash = sha256(_output);
+        if (
+            !IFunctionVerifier(verifier).verify(
+                inputHash,
+                outputHash,
+                _proof
+            )
+        ) {
+            revert InvalidProof();
+        }
+
+        // Default case
+        oracle[_functionId][inputHash] = outputHash;
+
+        if (_address != address(0)) {
```

Comment:
IMO we should have the callback functionality in a separate function rather than shove it in here